### PR TITLE
[feature](nereids) let user define right deep tree penalty by session variable

### PR DESCRIPTION
**Pull request by englefly**
# Proposed changes
it is hard for us to find a proper factor for all queries.
default is 0.7

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.12 seconds
 load time: 460 seconds
 storage size: 17179405369 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221107173810_clickbench_pr_41348.html

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 1]
```diff
@@ -191,6 +191,8 @@ public class SessionVariable implements Serializable, Writable {
     public static final String ENABLE_NEREIDS_RUNTIME_FILTER = "enable_nereids_runtime_filter";
 
     public static final String NEREIDS_STAR_SCHEMA_SUPPORT = "nereids_star_schema_support";
+
+    public static final String NEREIDS_PENALTY_FACTOR = "nereids_penalty_factor";
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:195]:
```suggestion
    public static final String NEREIDS_CBO_PENALTY_FACTOR = "nereids_cbo_penalty_factor";
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java [hunk 2]
```diff
@@ -80,7 +81,9 @@ public static double calculateCost(GroupExpression groupExpression) {
         CostEstimator costCalculator = new CostEstimator();
         CostEstimate costEstimate = groupExpression.getPlan().accept(costCalculator, planContext);
         groupExpression.setCostEstimate(costEstimate);
-        CostWeight costWeight = new CostWeight(CPU_WEIGHT, MEMORY_WEIGHT, NETWORK_WEIGHT, PENALTY_WEIGHT);
+        //CostWeight costWeight = new CostWeight(CPU_WEIGHT, MEMORY_WEIGHT, NETWORK_WEIGHT, PENALTY_WEIGHT);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java:84]: remove this line

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java [hunk 3]
```diff
@@ -51,14 +51,15 @@ public class CostCalculator {
...
-    static final double PENALTY_WEIGHT = 0.5;
+
+    ///**
+    // * Except stats information, there are some special criteria in doris.
+    // * For example, in hash join cluster, BE could build hash tables
+    // * in parallel for left deep tree. And hence, we need to punish right deep tree.
+    // * penalyWeight is the factor of punishment.
+    // * The punishment is denoted by stats.penalty.
+    // */
+    //static final double PENALTY_WEIGHT = 0.5;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java:62]: remove these code

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 4]
```diff
@@ -147,16 +147,33 @@ private static JoinEstimationResult estimateInnerJoin(PhysicalHashJoin join, Equ
...
+            }
         } else if (joinType == JoinType.RIGHT_SEMI_JOIN || joinType == JoinType.RIGHT_ANTI_JOIN) {
-            rowCount = rightStats.getRowCount();
+            double rightCount = rightStats.getRowCount();
+            double leftCount = leftStats.getRowCount();
+            if (join.getHashJoinConjuncts().isEmpty()) {
+                rowCount = joinType == JoinType.RIGHT_SEMI_JOIN ? rightCount : 0;
+            } else {
+                rowCount = estimateLeftSemiJoin(rightCount, leftCount);
+            }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:176]: already in another PR

**Review by morrySnow — APPROVED** [review]

**Pull request merged**